src/Core/Domain/Security/Provider/OpenIdProvider.php

```diff
@@ -0,0 +1,575 @@
+        }
+
+        if ($this->configuration->getTokenEndpoint() === null) {
+            throw OpenIdConfigurationException::missingTokenEndpoint();
+        }
+        if (
+            $this->configuration->getIntrospectionTokenEndpoint() === null
+            && $this->configuration->getUserInformationEndpoint() === null
+        ) {
+            throw OpenIdConfigurationException::missingInformationEndpoint();
+        }
+
+        $this->verifyThatClientIsAllowedToConnectOrFail($clientIp);
+
+        $this->sendRequestForConnectionTokenOrFail($authorizationCode);
+        if ($this->providerToken->isExpired() && $this->refreshToken->isExpired()) {
+            throw SSOAuthenticationException::tokensExpired(OpenIdConfiguration::NAME);
+        }
+        if ($this->configuration->getIntrospectionTokenEndpoint() !== null) {
+            $this->sendRequestForIntrospectionTokenOrFail();
+        }
+
+        $this->username = $this->getUsernameFromLoginClaim();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function getUser(): ?ContactInterface
+    {
+        $this->info('Searching user : ' . $this->username);
+        $user = $this->contactService->findByName($this->username);
+        if ($user === null) {
+            $user = $this->contactService->findByEmail($this->username);
+        }
+
+        return $user;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function refreshToken(AuthenticationTokens $authenticationTokens): AuthenticationTokens
+    {
+        if ($authenticationTokens->getProviderRefreshToken() === null) {
+            throw SSOAuthenticationException::noRefreshToken();
+        }
+        $this->info(
+            'Refreshing token using refresh token',
+            [
+                'refresh_token' => substr($authenticationTokens->getProviderRefreshToken()->getToken(), -10)
+            ]
+        );
+        // Define parameters for the request
+        $data = [
+            "grant_type" => "refresh_token",
+            "refresh_token" => $authenticationTokens->getProviderRefreshToken()->getToken(),
+            "scope" => !empty($this->configuration->getConnectionScopes())
+                ? implode(' ', $this->configuration->getConnectionScopes())
+                : null
+        ];
+        $headers = [
+            'Content-Type' => "application/x-www-form-urlencoded"
+        ];
+
+        // Define authentication type based on configuration
+        if ($this->configuration->getAuthenticationType() === OpenIdConfiguration::AUTHENTICATION_BASIC) {
+            $headers['Authorization'] = "Basic " . base64_encode(
+                $this->configuration->getClientId() . ":" . $this->configuration->getClientSecret()
+            );
+        } else {
+            $data["client_id"] = $this->configuration->getClientId();
+            $data["client_secret"] = $this->configuration->getClientSecret();
+        }
```

Comment:
this could be managed in a private method to avoid doing the same in `sendRequestForConnectionTokenOrFail`